File: src/token.cpp
#include "token.h"
#include <any>
#include <iostream>

namespace Jaql
{
    Token::Token(TokenType type, std::string lexeme, int line): 
        Token(type, lexeme, std::any{}, line)
    {};

    Token::Token(TokenType type, std::string lexeme, std::any literal, int line):
        type(type),
        lexeme(std::move(lexeme)),
        literal(literal),
        line(line)
    {};

    std::string Token::to_string() const
    {
        std::string output = "Type: " + std::to_string(static_cast<int>(type)) + ", Lexeme: " + lexeme ;
        if (type == TokenType::STRING or type == TokenType::NUMBER){
             output+= ", Literal: "; 
             output  += literal_to_string() ;
        }
        return output;
    };
    std::string Token::literal_to_string() const
    {
        switch (type) {
        case TokenType::STRING:
            return std::any_cast<std::string>(literal);
        case TokenType::NUMBER:
            return std::to_string(std::any_cast<double>(literal));
        default:
            return "";
        }
    }
};

File: src_py/token.py
from typing import Any

from src_py.token_type import TokenType


class Token:
    def __init__(self, type: TokenType, lexeme: str, line: int, literal: Any = None):
        self.type: TokenType = type
        self.lexeme: str = lexeme
        self.line: int = line
        self.literal: Any = literal

    def to_string(self):
        output = f"Type: {self.type}, Lexeme: {self.lexeme}"

        if self.type in (TokenType.STRING, TokenType.NUMBER):
            output += f", Literal: {self.literal_to_string()}"

        return output

    def literal_to_string(self):
        """
        Right now clearly not needed because of python casting, but will become
        needed for numerical printing
        """
        match self.type:
            case TokenType.STRING:
                return self.literal
            case TokenType.NUMBER:
                return self.literal
            case _:
                return ""
